src/target/riscv/riscv.c

```diff
@@ -5043,89 +5121,54 @@ int riscv_enumerate_triggers(struct target *target)
 	if (r->triggers_enumerated)
 		return ERROR_OK;
 
-	r->triggers_enumerated = true;	/* At the very least we tried. */
+	if (target->state != TARGET_HALTED) {
+		LOG_TARGET_ERROR(target, "Unable to enumerate triggers: target not halted.");
+		return ERROR_FAIL;
+	}
 
-	riscv_reg_t tselect;
-	int result = riscv_get_register(target, &tselect, GDB_REGNO_TSELECT);
+	riscv_reg_t orig_tselect;
+	int result = riscv_get_register(target, &orig_tselect, GDB_REGNO_TSELECT);
 	/* If tselect is not readable, the trigger module is likely not
-		* implemented. There are no triggers to enumerate then and no error
-		* should be thrown. */
+	 * implemented. */
 	if (result != ERROR_OK) {
-		LOG_TARGET_DEBUG(target, "Cannot access tselect register. "
-				"Assuming that triggers are not implemented.");
-		r->trigger_count = 0;
-		return ERROR_OK;
+		LOG_DEBUG("[%s] Cannot access tselect register. "
+				"Assuming that triggers are not implemented.", target_name(target));
+		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
 	}
 
-	for (unsigned int t = 0; t < RISCV_MAX_TRIGGERS; ++t) {
-		r->trigger_count = t;
-
-		/* If we can't write tselect, then this hart does not support triggers. */
-		if (riscv_set_register(target, GDB_REGNO_TSELECT, t) != ERROR_OK)
+	unsigned int t = 0;
+	unsigned int trigger_tinfo[ARRAY_SIZE(r->trigger_tinfo)];
```

Comment:
Here you can zero-initialize the whole array -- so that anyone looking at it via debugger does not see any random garbage in the unused array items.

```suggestion
	unsigned int trigger_tinfo[ARRAY_SIZE(r->trigger_tinfo)] = {0};
```

Reply:
Addressed.